extensions/data-plane-selector/store/cosmos/data-plane-instance-store-cosmos/src/main/java/org/eclipse/edc/connector/dataplane/selector/store/cosmos/CosmosDataPlaneInstanceStore.java

```diff
@@ -0,0 +1,94 @@
+ *
+ */
+
+package org.eclipse.edc.connector.dataplane.selector.store.cosmos;
+
+import dev.failsafe.RetryPolicy;
+import org.eclipse.edc.azure.cosmos.CosmosDbApi;
+import org.eclipse.edc.azure.cosmos.dialect.SqlStatement;
+import org.eclipse.edc.connector.dataplane.selector.spi.instance.DataPlaneInstance;
+import org.eclipse.edc.connector.dataplane.selector.spi.store.DataPlaneInstanceStore;
+import org.eclipse.edc.spi.persistence.EdcPersistenceException;
+import org.eclipse.edc.spi.types.TypeManager;
+
+import java.util.Collection;
+import java.util.stream.Stream;
+
+import static dev.failsafe.Failsafe.with;
+
+/**
+ * Implementation of the {@link DataPlaneInstanceStore} based on CosmosDB. This store implements simple write-through
+ * caching mechanics: read operations (e.g. findAll) always hit the cache, while write operations affect both the cache
+ * AND the database.
+ */
+public class CosmosDataPlaneInstanceStore implements DataPlaneInstanceStore {
+    private final CosmosDbApi cosmosDbApi;
+
+    private final TypeManager typeManager;
+
+    private final RetryPolicy<Object> retryPolicy;
+    private final String partitionKey;
+
+    public CosmosDataPlaneInstanceStore(CosmosDbApi cosmosDbApi, TypeManager typeManager,
+                                        RetryPolicy<Object> retryPolicy, String partitionKey) {
+        this.cosmosDbApi = cosmosDbApi;
+        this.typeManager = typeManager;
+        this.retryPolicy = retryPolicy;
+        this.partitionKey = partitionKey;
+    }
+
+    @Override
+    public void save(DataPlaneInstance instance) {
+        try {
+            insertOrUpdate(instance);
+        } catch (Exception exception) {
+            throw new EdcPersistenceException(exception);
+        }
+    }
+
+    @Override
+    public void saveAll(Collection<DataPlaneInstance> instances) {
+        try {
+            instances.forEach(this::insertOrUpdate);
+        } catch (Exception exception) {
+            throw new EdcPersistenceException(exception);
+        }
+    }
+
+    @Override
+    public DataPlaneInstance findById(String id) {
+        var dataSpaceInstance = with(retryPolicy).get(() -> cosmosDbApi.queryItemById(id));
+        return convert(dataSpaceInstance);
+    }
+
+    @Override
+    public Stream<DataPlaneInstance> getAll() {
+        var statement = new SqlStatement<>(DataPlaneInstanceDocument.class);
+        var query = statement.getQueryAsString();
+        var dataSpaceInstances = with(retryPolicy).get(() -> cosmosDbApi.queryItems(query));
+        return dataSpaceInstances.map(this::convert);
+    }
+
+    private void insertOrUpdate(DataPlaneInstance instance) {
+        var document = new DataPlaneInstanceDocument(instance, partitionKey);
+        with(retryPolicy).run(() -> cosmosDbApi.saveItem(document));
+    }
+
+    private DataPlaneInstance convert(Object object) {
+        var json = typeManager.writeValueAsString(object);
+        return json != null ?
```

Comment:
An `Optional` could be used to avoid null checking

Reply:
Got it